### PR TITLE
Federator NFT

### PR DESCRIPTION
**Pull request by RafilxTenfen**
### Add Federator for reading NFT cross Event

Besides some config files needed to work with typescript, this is the first service in the federator for the NFT bridge

### Description

- Modified cross event variable names
- Add `FederatorNFT.ts` to the federator directory

### How to Test
- Actually, it's not possible to test it yet, but you can ensure that the tests for the main bridge are working properly ;) 

- Run both `ganache-cli`
- Enter into the federator directory and run the integration test 

#### Case 1

1. Go to bridge directory
```shell
$~ cd bridge
```

2. Run the first ganache-cli
```shell
$~ npm run ganache
```

3. Open another shell and run the ganache mirror
```shell
$~ npm run ganache-mirror
```

4. Compile and deploy the contracts to the ganache chains
```shell
$~ npm run deployLocalIntegrationTest
```

5. Go to the Federator directory and run the test for the main bridge federator
```shell
$~ cd ../federator
$~ npm run test
```

__Expected Result__
- It should pass the tests
image

#### Bridge Directory `cd bridge`
- [x] Lint is clean `npm run lint`
- [x] Test is passing `npm run test`
- [x] Contracts are compiling `npm run compile`

#### Federator Directory `cd federator`
- [x] Lint is clean `npm run lint`
- [x] Test is passing `npm run test`
- [x] Typescript is compiling `npm run build`

### CODE REVIEW

File: federator/src/contracts/IBridgeNft.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  getAddress(): string {
+    return this.nftBridgeContract.options.address;
+  }
+
+  getProcessed(transactionHash: string): any {
+    return this.nftBridgeContract.methods.processed(transactionHash).call();
+  }
+
+  getVersion(): string {
+    return this.nftBridgeContract.methods.version();
```
Comment (pmprete) [inline federator/src/contracts/IBridgeNft.ts:25]: Add call to retrieve the version
```suggestion
    return this.nftBridgeContract.methods.version().call();
```

File: federator/src/contracts/IBridgeNft.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+export class IBridgeNft {
+  nftBridgeContract: any;
+
+  constructor(nftBridgeContract: any) {
+    this.nftBridgeContract = nftBridgeContract;
+  }
+
+  getFederation(): any {
+    return this.nftBridgeContract.methods.getFederation();
```
Comment (pmprete) [inline federator/src/contracts/IBridgeNft.ts:9]: Add .call() to invoke the method
```suggestion
    return this.nftBridgeContract.methods.getFederation();
```

File: federator/src/contracts/IBridgeNft.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+export class IBridgeNft {
+  nftBridgeContract: any;
+
+  constructor(nftBridgeContract: any) {
+    this.nftBridgeContract = nftBridgeContract;
+  }
+
+  getFederation(): any {
```
Comment (pmprete) [inline federator/src/contracts/IBridgeNft.ts:8]: Try to add types when you can, specially the Promise if they are async.
All addresses are returned as strings, numbers are returned as BigNumbers from BN.js
Hashes are returned as strings
```suggestion
  getFederation(): Promise<string> {
```

File: federator/src/contracts/IBridgeNft.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  getPastEvents(eventName: string, options: any): any {
+    return this.nftBridgeContract.getPastEvents(eventName, options);
+  }
+
+  getAddress(): string {
+    return this.nftBridgeContract.options.address;
+  }
+
+  getProcessed(transactionHash: string): any {
```
Comment (pmprete) [inline federator/src/contracts/IBridgeNft.ts:20]:
```suggestion
  getProcessed(transactionHash: string): Promise<boolean> {
```

File: federator/src/contracts/IBridgeNft.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    return this.nftBridgeContract.methods.processed(transactionHash).call();
+  }
+
+  getVersion(): string {
```
Comment (pmprete) [inline federator/src/contracts/IBridgeNft.ts:24]:
```suggestion
  getVersion(): Promise<string> {
```

File: federator/src/lib/FederatorNFT.ts [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+  public transactionSender: TransactionSender;
+  public bridgeFactory: BridgeFactory;
+  public federationFactory: FederationFactory;
+  public nftConfirmationsBlocks: number;
+  private federatorContract: import('../contracts/IFederationV2');
+
+  constructor(config: Config, logger: Logger, Web3 = web3) {
+    this.config = config;
+    this.logger = logger;
+    this.nftConfirmationsBlocks = config.nftConfirmations;
```
Comment (pmprete) [inline federator/src/lib/FederatorNFT.ts:29]: In this cases it's easier a ternary if or when checking for null, empty or undefined just use ??
```suggestion
    this.nftConfirmationsBlocks = config.nftConfirmations ?? NFT_CONFIRMATION_BLOCKS;
```

File: federator/src/main.js [hunk 7]
```diff
@@ -49,6 +60,16 @@ async function run() {
     try {
         await mainFederator.run();
         await sideFederator.run();
+
+        if (config.mainchain.nftBridge !== null) {
```
Comment (pmprete) [inline federator/src/main.js:64]: add a check that if config.mainchain.nftBridge and config.sideChain.nftBridge is not defined throw an error and viceversa

**Review by pmprete — APPROVED** [review]
Except some coments, looks good!

File: federator/src/main.js (hunk 7 above)
Comment (RafilxTenfen) [reply federator/src/main.js]: done at https://github.com/rsksmart/tokenbridge/pull/257/commits/960a77254356ad684efe45071b291b518e0266f8

File: federator/src/lib/FederatorNFT.ts (hunk 6 above)
Comment (RafilxTenfen) [reply federator/src/lib/FederatorNFT.ts]: done at https://github.com/rsksmart/tokenbridge/pull/257/commits/960a77254356ad684efe45071b291b518e0266f8

File: federator/src/contracts/IBridgeNft.ts (hunk 5 above)
Comment (RafilxTenfen) [reply federator/src/contracts/IBridgeNft.ts]: done at https://github.com/rsksmart/tokenbridge/pull/257/commits/960a77254356ad684efe45071b291b518e0266f8

File: federator/src/contracts/IBridgeNft.ts (hunk 4 above)
Comment (RafilxTenfen) [reply federator/src/contracts/IBridgeNft.ts]: done at https://github.com/rsksmart/tokenbridge/pull/257/commits/960a77254356ad684efe45071b291b518e0266f8

File: federator/src/contracts/IBridgeNft.ts (hunk 3 above)
Comment (RafilxTenfen) [reply federator/src/contracts/IBridgeNft.ts]: done at https://github.com/rsksmart/tokenbridge/pull/257/commits/960a77254356ad684efe45071b291b518e0266f8

File: federator/src/contracts/IBridgeNft.ts (hunk 2 above)
Comment (RafilxTenfen) [reply federator/src/contracts/IBridgeNft.ts]: done at https://github.com/rsksmart/tokenbridge/pull/257/commits/960a77254356ad684efe45071b291b518e0266f8

File: federator/src/contracts/IBridgeNft.ts (hunk 1 above)
Comment (RafilxTenfen) [reply federator/src/contracts/IBridgeNft.ts]: done at https://github.com/rsksmart/tokenbridge/pull/257/commits/960a77254356ad684efe45071b291b518e0266f8

File: federator/src/contracts/FederationFactory.js [hunk 8]
```diff
@@ -1,6 +1,7 @@
 const abiFederationOld = require('../../../bridge/abi/Federation_old.json');
 const abiFederationNew = require('../../../bridge/abi/Federation.json');
 const abiBridge = require('../../../bridge/abi/Bridge.json');
+const abiNftBridge = require('../../../bridge/abi/Bridge.json'); // change to bridgeNft after deploy
```
[the remark this answers is not in the document]
Comment (RafilxTenfen) [reply federator/src/contracts/FederationFactory.js]: done at https://github.com/rsksmart/tokenbridge/pull/257/commits/960a77254356ad684efe45071b291b518e0266f8

**Pull request merged**